docs/rfc/rfc-022-semi-permanent-testnet.md

```diff
@@ -0,0 +1,202 @@
+and the expected outcomes from following this roadmap.
+
+## Background
+
+The author's understanding -- which is limited as a new contributor to the Tendermint project --
+is that Tendermint development has been largely treated as a library for other projects to consume.
+Of course effort has been spent on unit tests, end-to-end tests, and integration tests.
+But whether developing a library or an application,
+there is no substitute for putting the software under a production-like load.
+
+First, there are classes of bugs that are unrealistic to discover in environments
+that do not resemble production.
+But perhaps more importantly, there are "operational features" that are best designed
+by the authors of a given piece of software.
+For instance, does the software have sufficient observability built-in?
+Are the reported metrics useful?
+Are the log messages clear and sufficiently detailed, without being too noisy?
+
+Furthermore, if the library authors are not only building --
+but also maintaining and operating -- an application built on top of their library,
+the authors will have a greatly increased confidence that their library's API
+is appropriate for other application authors.
+
+Once the decision has been made to run and operate a service,
+one of the next strategic questions is that of deploying said service.
+The author strongly holds the opinion that, when possible,
+a continuous delivery model offers the most compelling set of advantages:
+- The code on a particular branch (likely `main` or `master`) is exactly what is,
+  or what will very soon be, running in production
+- There are no manual steps involved in deploying -- other than merging your pull request,
+  which you had to do anyway
+- A bug discovered in production can be rapidly confirmed as fixed in production
+
+In summary, if the tendermint authors build, maintain, and continuously deliver an application
+intended to serve as a long-lived testnet, they will be able to state with confidence:
+- We operate the software in a production-like environment and we have observed it to be
+  stable and performant to our requirements
+- We have discovered issues in production before any external parties have consumed our software,
+  and we have addressed said issues
+- We have successfully used the observability tooling built into our software
+  (perhaps in conjunction with other off-the-shelf tooling)
+  to diagnose and debug issues in production
+
+## Discussion
+
+The Discussion Section proposes a variety of aspects of maintaining a testnet for Tendermint.
+
+### Number of testnets
+
+There should probably be one testnet per maintained branch of Tendermint,
+i.e. one for the `main` branch
+and one per `v0.N.x` branch that the authors maintain.
+
+We may eventually discover that there is good reason to run more than one testnet for a branch,
+perhaps due to a significant configuration variation.
```

Comment:
Or a feature that was introduced initially in a branch, e.g. PBTS or ABCI++.

In fact, Willian has run some experiments with PBTS to attest that it was not messing out blocks, before merging the new code to master.

Reply:
Updated in 3e1d75805.